design-proposals/kubebmc.md

```diff
@@ -0,0 +1,156 @@
+
+## Goals
+
+- Provide essential BMC functionalities such as (rebooting, changing boot devices, and mounting virtual media) via IPMI and Redfish protocols against VMs backed by KubeVirt
+- Expose service endpoints for clients to access either in-cluster or externally
+- Enable deployment tools, e.g., [Metal3](https://metal3.io/), [Tinkerbell](https://tinkerbell.org/), [Bifrost](https://opendev.org/openstack/bifrost/), etc., to install software on VMs without adapting to the virtual environment by modifying their source code
+
+## Non Goals
+
+- Providing BMC functionalities for bare-metal machines
+
+## Definition of Users
+
+- Users who want to do out-of-band management to the VMs within a KubeVirt environment directly using bare-metal-like methods, e.g., IPMI or Redfish
+- Users who want to install software like OpenShift or OpenStack on VMs locally within a KubeVirt environment using existing deployment tools
+
+## User Stories
+
+- As a developer, I want to deploy a multi-node OpenShift cluster using its native installer in KubeVirt VMs.
+- As a tester, I need to quickly spin up and destroy OpenStack clusters in a virtual environment for testing purposes.
+
+## Repos
+
+No existing repos under the KubeVirt organization will be impacted. KubeBMC augments the KubeVirt ecosystem as an add-on.
+
+# Design
+
+KubeBMC will act as a bridge between KubeVirt VMs and the BMC functionalities, using Kubernetes services to expose these capabilities. KubeBMC consists of two components:
+
+- **kubebmc-controller**: A typical Kubernetes controller reconciles the KubeBMC, VirtualMachine, and Service objects
+- **kbmc**: A BMC simulator for serving IPMI and Redfish endpoints and translating the incoming requests to native Kubernetes API calls
+
+Each VirtualMachine will have its own KubeBMC object created by `kubebmc-controller` during the VM creation. The controller will further process the KubeBMC object to complete the BMC scaffold. This includes a BMC emulator Pod, i.e., `kbmc`, and its Service object working together to provide the BMC functionalities to various clients over the network.
+
+Inside the BMC emulator, the IPMI/Redfish requests are translated to native Kubernetes API calls and sent to the API server. It's essentially a combination of IPMI/HTTP server and Kubernetes client.
+
+The following graph illustrates the entire KubeBMC workflow:
+
+```mermaid
+flowchart LR
+    controller["kubebmc-controller"]
+    cr["kubebmc CR"]
+    kbmc-pod["kbmc Pod"]
+    kbmc-svc["kbmc Service"]
+    controller-.->|watches|cr
+    cr-.->|owns|kbmc-svc
+    cr-.->|owns|kbmc-pod
+    client--->|IPMI|kbmc-svc
+    kbmc-svc-->kbmc-pod
+    kbmc-pod-->|HTTP|apiserver
+    apiserver-->|modifies|vm
+    vm-->|creates|vmi
+```
+
+It's also possible for VMs being created to opt out of such a functionality. This could be done by annotating the VirtualMachine object with a specific key-value pair.
```

Comment:
I would prefer to make this opt-in instead of opt-out. Can the automatic creation of BMCs be controlled e.g. by a filter set on the controllers CRD? For a start we could keep creation of BMC CRDs a manual task?

Reply:
It could be done like that. There's a VM reconciler ensuring the 1:1 mapping between VirtualMachine and KubeBMC objects. By the "filter set" you mentioned above, did you mean that a one-to-many relationship is preferable?

Reply:
I meant a filter with which you can select the VMs a BMC is created for, e.g. a label selector. The filter could be configured in the CRD/CR of the BMC controller.

Reply:
We could, but the current implementation is a 1:1 mapping with a namespaced name reference between the BMC and VM. If a filter-like label selector is used instead, it implies a BMC could map to more than one VM. We need to multiplex them using different ports. That's a different approach.

Back to the original topic, do you think adding an annotation to the VM CRD to opt-in is a good idea?

Reply:
> For a start we could keep creation of BMC CRDs a manual task?

This sounds most reasonable to me to start with. No "magic" annotations, less complexity and surprises at the cost of slightly more work for the user creating the custom resources.

Reply:
Did I understand correctly, that in your proposal `kubebmc-controller` is supposed to create a BMC CRD for every VM? I was thinking we should make `kubebmc-controller` configurable by adding a label selector to its configuration, so the label selector controlls for which VMs a BMC CRD will be created. I would choose labels, since Kubernetes intends labels to be used to identify objects. 

As suggested and backed by @lentzi90 we could implement that at a later stage though.

Reply:
+1 to default to manual CR creation to begin with and adding label selectors etc later.

Reply:
> Did I understand correctly, that in your proposal kubebmc-controller is supposed to create a BMC CRD for every VM?

Yes, the `kubebmc-controller` in the current implementation will create a BMC CRD (and thus a pair of Pod and Service) for each VM, so it's a one-to-one mapping. If we change the specifier from a namespaced name to a label selector, does that mean a one-to-many mapping is preferable?